### PR TITLE
Alessandro/feature/video filtering

### PR DESCRIPTION
**Pull request by alessandropalla**
Video filtering demo

### CODE REVIEW

File: wadas/ai/object_tracker.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+            # Update classification
+            class_filters = self.trackers[obj_id][1]
+            smoothed_probs = {
+                cls: class_filters[cls].update([p])[0] for cls, p in class_probs.items()
+            }
+
+            # Build classification result
+            logits = np.array(list(smoothed_probs.values()))
+            labels = list(smoothed_probs.keys())
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:256]: labels = tuple(smoothed_probs)

File: wadas/ai/object_tracker.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            classification_result = labels[np.argmax(logits)], max(logits)
+
+            updated_tracks.append(
+                {"id": obj_id, "classification": classification_result, "xyxy": smoothed_pos}
+            )
+
+        # Increment missed count for unmatched trackers
+        for obj_id in self.trackers.keys():
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:264]: Redundant .keys()

File: wadas/ai/object_tracker.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                    position_filter = self.trackers[obj_id][0]
+                    smoothed_pos = position_filter.update(position_filter.x[:2])
+
+                    # Convert smoothed_pos to xyxy
+                    h, w = position_filter.extra["h"], position_filter.extra["w"]
+                    smoothed_pos = [
+                        smoothed_pos[0] - w / 2,
+                        smoothed_pos[1] - h / 2,
+                        smoothed_pos[0] + w / 2,
+                        smoothed_pos[1] + h / 2,
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:282]: //?

File: wadas/ai/object_tracker.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                    ]
+
+                    # Convert to int
+                    smoothed_pos = [int(p) for p in smoothed_pos]
+
+                    # Update classification
+                    class_filters = self.trackers[obj_id][1]
+                    smoothed_probs = {
+                        cls: class_filters[cls].update([class_filters[cls].x[0]])[0]
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:291]: If class_filters[cls] is dictionary, .update() always gives back None.

File: wadas/ai/object_tracker.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+                        for cls in class_filters.keys()
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:292]: Redundant .keys() ?

File: wadas/ai/object_tracker.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+                    }
+
+                    # Build classification result
+                    logits = np.array(list(smoothed_probs.values()))
+                    labels = list(smoothed_probs.keys())
```
Comment (cdimauro) [inline wadas/ai/object_tracker.py:297]: labels = tuple(smoothed_probs)

File: wadas/ai/object_tracker.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        """
+        Compute the centroid of a bounding box.
+        Args:
+            xyxy (list[int]): A list of four integers representing the coordinates
+                              of the bounding box in the format [x1, y1, x2, y2].
+        Returns:
+            tuple[int]: A tuple containing the x and y coordinates of the centroid.
+        """
+
+        return (xyxy[0] + xyxy[2]) / 2, (xyxy[1] + xyxy[3]) / 2
```
[the remark this answers is not in the document]
Comment (alessandropalla) [reply wadas/ai/object_tracker.py:129]: Actually, this is a feature, the centroid needs to be float

File: wadas/ai/object_tracker.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        if not self.trackers:
+            return {i: self.next_id + i for i in range(len(detections))}  # Assign new IDs
+
+        # Compute IoU matrix
+        track_ids = list(self.trackers.keys())
+        track_boxes = []
+        for t in track_ids:
+            h, w = self.trackers[t][0].extra["h"], self.trackers[t][0].extra["w"]
+            x, y, _, __ = self.trackers[t][0].x
+            xyxy = [x - w / 2, y - h / 2, x + w / 2, y + h / 2]
```
[the remark this answers is not in the document]
Comment (alessandropalla) [reply wadas/ai/object_tracker.py:157]: Actually, this is a feature, the centroid needs to be float

File: wadas/ai/object_tracker.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+            # Update position
+            position_filter = self.trackers[obj_id][0]
+            smoothed_pos = position_filter.update([x, y])
+
+            # Convert smoothed_pos to xyxy
+            smoothed_pos = [
+                smoothed_pos[0] - w / 2,
+                smoothed_pos[1] - h / 2,
+                smoothed_pos[0] + w / 2,
+                smoothed_pos[1] + h / 2,
```
[the remark this answers is not in the document]
Comment (alessandropalla) [reply wadas/ai/object_tracker.py:242]: as before

File: wadas/ai/object_tracker.py (hunk 3 above)
Comment (alessandropalla) [reply wadas/ai/object_tracker.py:282]: as before

File: wadas/ai/object_tracker.py (hunk 4 above)
Comment (alessandropalla) [reply wadas/ai/object_tracker.py:291]: It is not a python dict, is a Kalman Filter

File: wadas/ai/object_tracker.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
+import numpy as np
```
Comment (stefanodellosa-personal) [inline wadas/ai/object_tracker.py:1]: License header

File: wadas/domain/ai_model.py [hunk 11]
```diff
@@ -159,6 +156,19 @@ def process_video(self, video_path, save_detection_image: bool):
...
+
+    def process_video(self, video_path, save_detection_image: bool):
+        """Method to run detection model on provided video."""
+
+        logger.debug("Selected detection device: %s", AiModel.detection_device)
+
+        logger.info("Running detection on video %s ...", video_path)
+        video_filename = os.path.basename(video_path)
+
+        for frame, frame_count in self.get_video_frames(video_path):
```
Comment (stefanodellosa-personal) [inline wadas/domain/ai_model.py:170]: perhaps here we need a way to interrupt video processing if stop button is triggered?
